### PR TITLE
CI / Local dev refresh and Helm2->3 transition

### PR DESCRIPTION
**Pull request by consideRatio**
With a easy to understand CI system without tangled configuration that influence too many things at once, I hope that it will help myself and others to make contributions to this project in the long term. I wanted to update this step by step towards being easier to understand, maintain, and update it with tricks learned over time.

## Work to be done
- [x] Inline comment files on how they are used
- [x] Collect testing config into testing folder
- [x] Test with helm3 instead of helm2
- [x] Create a dedicated publishing step
- [x] Minimize rebuilds by configuring chartpress as well as possible
  - [x] Develop the rebuildOnContextPathChanges configuration option in chartpress (https://github.com/jupyterhub/chartpress/pull/98 - __Review help wanted!__)
  - [x] Release chartpress 0.7.0 (https://github.com/jupyterhub/chartpress/pull/99)
  - [x] Bump to use chartpress 0.7.0 in this repo

## Review notes
- __No tooling changes__. We still use `minikube` and I have not introduced new tooling (except the helm v2 to v3 migration).
- __No changes to BinderHub or Helm chart itself__. This PR only impacts the CI system local development instructions, and the guide with regards to now having instructions on how to use Helm 3 rather than Helm 2.
- __CONTRIBUTING.md would benefit from your review.__ I have improved them especially with regards to anyone wanting to develop the Helm chart in full rather than a BinderHub running locally interacting with a JupyterHub running in k8s.

## Positive outcomes
1. __CI refactoring reduced complexity__. This will make it easier to maintain for current maintainers, but for me, the most important part is that it can attract more maintenance contributions which makes the project more sustainable in the long term.
1. __Helm2 to Helm3 migration complete.__
1. __Helm3 based template and rendered template validation.__ `helm3 template --validate` both renders the templates but also validates them as valid k8s resources against a k8s api-server. It is a more heavy weight option to using `kubeval` and such tools as it requires a k8s api-server to speak with.
1. __Faster CI jobs__. With the chartpress feature developed, we can avoid triggering rebuilds of the BinderHub docker image for the Helm chart all the time, which can save us multiple minutes in build time.

### CODE REVIEW

File: .gitattributes [hunk 1]
```diff
@@ -1,2 +1 @@
-secrets/** filter=git-crypt diff=git-crypt
```
Comment (sgibson91) [inline .gitattributes:1]: Why has this been removed? Have the secrets been migrated elsewhere?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -43,233 +49,253 @@ uses the [reStructuredText markup language](http://www.sphinx-doc.org/en/master/
...
    npm install
    ```
 
-1. Create the JS and CSS bundles with:
+1. Create the JS and CSS bundles BinderHub serves as a webserver to visitors.
+   Repeat this step if you change the source JS or CSS.
 
    ```bash
    npm run webpack
    ```
```
Comment (sgibson91) [inline CONTRIBUTING.md:83]:
```suggestion
1. Create the JS and CSS bundles BinderHub serves as a webserver to visitors.

   ```bash
   npm run webpack:watch
   ```
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -43,233 +49,253 @@ uses the [reStructuredText markup language](http://www.sphinx-doc.org/en/master/
...
-### One-time installation
+## Develop Helm chart
 
-To run the full BinderHub it needs to have access to a kubernetes cluster
-with a JupyterHub installed on it. This is what we will setup in this section.
-All the steps are given as command-line commands for Ubuntu systems. They are
-used as a common denominator that can be translated into the correct commands
-on your local system.
+This requires `helm` and a functional Kubernetes cluster. Please do preliminary
+Kubernetes setup if you haven't already before continuing here.
```
Comment (sgibson91) [inline CONTRIBUTING.md:163]:
```suggestion
This requires `helm` and a functional Kubernetes cluster. Please do
[preliminary Kubernetes setup](#kubernetes-setup) if you haven't already
before continuing here.
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -43,233 +49,253 @@ uses the [reStructuredText markup language](http://www.sphinx-doc.org/en/master/
...
 
 To learn how to set yourself with a BinderHub development environment that
 lets you modify the builder and launcher refer to
 [Kubernetes integration changes](#Kubernetes-integration-changes).
 
+## Develop Kubernetes integration
 
-## Kubernetes integration changes
+This requires `helm` and a functional Kubernetes cluster. Please do preliminary
+Kubernetes setup if you haven't already before continuing here.
```
Comment (sgibson91) [inline CONTRIBUTING.md:105]:
```suggestion
This requires `helm` and a functional Kubernetes cluster. Please do
[preliminary Kubernetes setup](#kubernetes-setup) if you haven't already
before continuing here.
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -43,233 +49,253 @@ uses the [reStructuredText markup language](http://www.sphinx-doc.org/en/master/
...
+1. To cleanup the JupyterHub Helm chart you have installed in Kubernetes...
 
-You can now access your BinderHub at: `http://192.168.99.100:30901`. If your
-minikube instance has a different IP use `minikube ip` to find it. You will
-have to use that IP in two places. Add `--set config.BinderHub.hub_url: http://$IP:30902`
-to your `helm install` command and access your BinderHub at `http://$IP:30901`.
-Replace `$IP` with the output of `minikube ip`.
+   ```bash
+   helm delete binderhub-test
+   ```
```
Comment (sgibson91) [inline CONTRIBUTING.md:153]:
```suggestion

### Cleanup resources

```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -43,233 +49,253 @@ uses the [reStructuredText markup language](http://www.sphinx-doc.org/en/master/
...
+      --set config.BinderHub.hub_url=http://$(minikube ip):30902 \
+      --set config.BinderHub.access_token=$GITHUB_ACCESS_TOKEN
+
+   echo "BinderHub inside the Minikube based Kubernetes cluster is starting up at http://$(minikube ip):30901"
+   ```
+
+1. Congratulations, you can now make changes and run the step above again to see
+   how they influence the deployment. You can also start running `pytest` tests
+   to verify the Deployment functions as it should.
+
```
Comment (sgibson91) [inline CONTRIBUTING.md:216]:
```suggestion

### Cleanup resources

```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -3,24 +3,30 @@
...
-the [user interface](#user-innterface-changes), or the
-[kubernetes integration](#Kubernetes-integration-changes).
+Depending on what you want to develop, you can setup BinderHub in different ways.
+- [Develop documentation](#).
+- [Develop user interface](#). A BinderHub webserver is running locally and
+  JupyterHub is mocked, this setup doesn't involve Kubernetes.
+- [Develop Kubernetes integration](#). A BinderHub webserver is running locally,
+  and JupyterHub is installed in a Kubernetes cluster.
+- [Develop Helm chart](#) - The BinderHub Helm chart with JupyterHub as a
+  dependency is installed in a Kubernetes cluster.
```
Comment (sgibson91) [inline CONTRIBUTING.md:13]:
```suggestion
  dependency is installed in a Kubernetes cluster.
 
 This document also contains information on [how to run tests](#running-tests) and
 [common maintainer tasks](#common-maintainer-tasks).
```

File: doc/zero-to-binderhub/setup-prerequisites.rst [hunk 8]
```diff
@@ -3,152 +3,62 @@
...
-
-
-.. note::
+Helm packages are called *charts*. We will be installing and managing JupyterHub
+on our Kubernetes cluster using a Helm chart.
 
-   These instructions are for Helm 2.
-   Helm 3 includes several major breaking changes and is not yet officially
-   supported
+A Helm *chart* are mostly Helm *templates* and default *values* that are used to
```
Comment (sgibson91) [inline doc/zero-to-binderhub/setup-prerequisites.rst:29]:
```suggestion
A Helm *chart* is mostly Helm *templates* and default *values* that are used to
```

Alternatively, this sentence could start _"Helm charts are mostly..."_

File: doc/zero-to-binderhub/setup-prerequisites.rst [hunk 9]
```diff
@@ -3,152 +3,62 @@
...
+render the templates into valid k8s resources. Each installation of a chart is
+called a *releases*, and each version of the release is called a *revision*.
```
Comment (sgibson91) [inline doc/zero-to-binderhub/setup-prerequisites.rst:31]:
```suggestion
called a *release*, and each version of the release is called a *revision*.
```

File: doc/zero-to-binderhub/setup-prerequisites.rst [hunk 10]
```diff
@@ -3,152 +3,62 @@
...
-
-First, deploy a Kubernetes cluster by following the
-`instructions in the Zero to JupyterHub guide <https://zero-to-jupyterhub.readthedocs.io/en/latest/create-k8s-cluster.html>`_.
+First, deploy a Kubernetes cluster by following the `instructions in the Zero to
+JupyterHub guide
+<https://zero-to-jupyterhub.readthedocs.io/en/latest/kubernetes/setup-kubernetes.html>`_.
 When you're done, move on to the next section.
 
-
 Installing Helm
```
Comment (sgibson91) [inline doc/zero-to-binderhub/setup-prerequisites.rst:21]: For the installing Helm section, why don't we also just link to the z2jh docs in order to minimise replication and number of places to maintain?

**Review by sgibson91 — COMMENTED** [review]
Thank you for this work @consideRatio! ✨ I have just left a few comments but this looks good to me :)

File: .gitattributes (hunk 1 above)
Comment (consideRatio) [reply .gitattributes:1]: This is an old remnant for when git-crypt was used in the binderhub repo to encrypt the SSH key for helm chart publishing or similar.

File: CONTRIBUTING.md (hunk 7 above)
Comment (consideRatio) [inline CONTRIBUTING.md:13]:
```suggestion
- [Develop documentation](#develop-documentation).
- [Develop user interface](#develop-user-interface). A BinderHub webserver is running locally and
  JupyterHub is mocked, this setup doesn't involve Kubernetes.
- [Develop Kubernetes integration](#develop-kubernetes-integration). A BinderHub webserver is running locally,
  and JupyterHub is installed in a Kubernetes cluster.
- [Develop Helm chart](#develop-helm-chart) - The BinderHub Helm chart with JupyterHub as a
  dependency is installed in a Kubernetes cluster.
```

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -43,233 +49,251 @@ uses the [reStructuredText markup language](http://www.sphinx-doc.org/en/master/
...
-1. build the helm chart `cd helm-chart && chartpress && cd ..`
-1. install the BinderHub chart with
-```
-helm install \
-  --name binder-test \
-  --namespace binder-test-helm \
-  helm-chart/binderhub \
-  -f helm-chart/minikube-binder.yaml
-```
+1. To cleanup the JupyterHub Helm chart you have installed in Kubernetes...
```
Comment (consideRatio) [inline CONTRIBUTING.md:146]:
```suggestion

### Cleanup resources

1. To cleanup the JupyterHub Helm chart you have installed in Kubernetes...
```

File: doc/zero-to-binderhub/setup-prerequisites.rst (hunk 10 above)
Comment (consideRatio) [reply doc/zero-to-binderhub/setup-prerequisites.rst:21]: Installing Helm 3 is trivial, only a binary needs to be installed at this point, so ensuring the documentation is in sync with z2jh is an about as complicated maintenance I think. But we if anyone wants to do this later I'd be fine with it.

**Comment by consideRatio** [comment]
@sgibson91 :heart: :tada: thank you for the review!!

**Comment by consideRatio** [comment]
I am not looking to many any more changes to this PR at this point, and it is ready for a final review.

Merging this gives us faster CI + faster upgrades by not replacing the docker image as often for no reason.

**Comment by minrk** [comment]
Looks nice. Thanks!

**Comment by consideRatio** [comment]
Chartpress 0.7.0 ability to not trigger a build of the images when not needed demonstrated to work in the CI system now.

image

**Pull request merged**